sql: tableoid column causes invalid memory access

[(roachtest).sqlsmith/setup=rand-tables/setting=no-ddl failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2231078&tab=buildLog) on [master@8b91062f9351d18f9104aff567cb152df162021e](https://github.com/cockroachdb/cockroach/commits/8b91062f9351d18f9104aff567cb152df162021e):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-ddl/run_1
	sqlsmith.go:172,sqlsmith.go:202,test_runner.go:754: error: pq: internal error: runtime error: invalid memory address or nil pointer dereference
		stmt:
		SELECT
			tab_15.col2_2 AS col_33,
			3.301392334012675018E+23:::DECIMAL AS col_34,
			737396787:::OID AS col_35,
			tab_15.col2_2 AS col_36
		FROM
			defaultdb.public.table2@[0] AS tab_15,
			defaultdb.public.table1@table1_col1_0_key AS tab_16
			JOIN defaultdb.public.table2@[0] AS tab_17 ON
					(tab_16.tableoid) = (tab_17.col1_0) AND (tab_16.col1_0) = (tab_17.col1_0),
			defaultdb.public.table2@table2_col1_0_idx AS tab_18;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-ddl](https://teamcity.cockroachdb.com/viewLog.html?buildId=2231078&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-ddl)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-ddl.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/logictest/testdata/logic_test/system_columns
+++ b/pkg/sql/logictest/testdata/logic_test/system_columns
@@ -186,5 +186,15 @@ SELECT tableoid, crdb_internal_mvcc_timestamp IS NOT NULL FROM tab1
 56 true
 56 true
 
+# We should be able to index join against the primary key to get the tableoid column.
+statement ok
+CREATE TABLE tab3 (x INT, INDEX i (x));
+INSERT INTO tab3 VALUES (1)
+
+query OI
+SELECT tableoid, x FROM tab3@i WHERE x = 1
+----
+58 1
+
 statement error pq: column name "tableoid" conflicts with a system column name
 CREATE TABLE bad (tableoid int)

--- a/pkg/sql/opt_catalog.go
+++ b/pkg/sql/opt_catalog.go
@@ -937,8 +937,12 @@ func (ot *optTable) getColDesc(i int) *descpb.ColumnDescriptor {
 	if i < len(ot.desc.DeletableColumns()) {
 		return &ot.desc.DeletableColumns()[i]
 	}
-	if ot.columns[i].ColID() == cat.StableID(colinfo.MVCCTimestampColumnID) {
-		return &colinfo.MVCCTimestampColumnDesc
+	// Check if the column matches any registered system columns.
+	for j := range colinfo.AllSystemColumnDescs {
+		colDesc := &colinfo.AllSystemColumnDescs[j]
+		if descpb.ColumnID(ot.columns[i].ColID()) == colDesc.ID {
+			return colDesc
+		}
 	}
 	return nil
 }

